Added Hilbert-Schmidt scalar product (CPU)

BalintKoczor opened the pull request:
New feature: Hilbert-Schmidt scalar product between two density operators

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -2447,6 +2447,25 @@ void applyTwoQubitKrausMap(Qureg qureg, int target1, int target2, ComplexMatrix4
+ *      if either \p a or \p b are not density matrices,
+ *      or if \p a and \p have mismatching dimensions.
+ */
+qreal calcHilbertSchmidtScalarProduct(Qureg a, Qureg b);
```
If the Hilbert Schmidt scalar product is a natural choice of inner product for density matrices, i.e. one analogous to `<a|b>`, then perhaps it's better to "overload" `calcInnerProduct`, and i.e. make this method `densmatr_calcInnerProduct`.

That way, user's code for processing statevectors is easily adapted to density matrices to e.g. introduce noise

BalintKoczor replied:
fixed this in commits 7af7983c750036da33a3074eb7462164c66e5368 and 8de0d6f2e62f1a2f150a7b26ae5ad13e28b6b53b

TysonRayJones replied:
(note I think the above response was meant for the previous QuEST.h changes; those commits do not address this review point)

BalintKoczor replied:
Sorry, that comment was ment to be for the above QuEST.h changes.

I updated the calcInnerProduct(...) function such that it now accepts density matrices and returns their Hilbert-Schmidt scalar product. See commits 1ea8ceb28a4a7b3a8025ce92b9129b894100911 
6c9f711d396ad704ef0aaaed8e9d04b80b19685d and 691f66020ba94a8149e7abc79619de06b4fae1b0

TysonRayJones commented on QuEST/src/CPU/QuEST_cpu.c:
```diff
@@ -942,6 +942,40 @@ qreal densmatr_calcHilbertSchmidtDistanceSquaredLocal(Qureg a, Qureg b) {
+# ifdef _OPENMP
+# pragma omp parallel \
+    shared    (aRe,aIm, bRe,bIm, numAmps) \
+    private   (index, traceIm) \
```
This looks bugged; `traceIm` can't be OpenMP `private` yet be accumulated in the loop. It seems `traceIm` was also meant to be in `reduction`

BalintKoczor replied:
removed this in bb79977ff0cddaf55527bb0dc1221e993bd466ef

TysonRayJones commented on QuEST/src/CPU/QuEST_cpu.c:
```diff
@@ -942,6 +942,40 @@ qreal densmatr_calcHilbertSchmidtDistanceSquaredLocal(Qureg a, Qureg b) {
+        }
+    }
+    
+    if(abs(traceIm) > REAL_EPS) printf("ERROR: too large imag. part of calcHilbertSchmidtScalarProduct");
```
I think `traceIm` shouldn't even be computed; a pre-condition is that density matrices `a` and `b` are Hermitian to numerical precision, and ergo `traceIm` should be zero. Users are warned in the doc for using Hermitivity-breaking functions, and this (outside of `validation`) isn't the right place to "throw" an error.

BalintKoczor replied:
Sure, this part of the code was irrelevant -- I removed it in bb79977ff0cddaf55527bb0dc1221e993bd466ef

BalintKoczor commented on QuEST/include/QuEST.h:
```diff
@@ -2447,6 +2447,25 @@ void applyTwoQubitKrausMap(Qureg qureg, int target1, int target2, ComplexMatrix4
+ * \f]
+ *
+ * @param[in] a a density matrix
+ * @param[in] b an equally-sized density matrix
```
fixed this in commits 7af7983c750036da33a3074eb7462164c66e5368 and 8de0d6f2e62f1a2f150a7b26ae5ad13e28b6b53b

TysonRayJones commented:
You've commited your cmake build files in bb79977ff0cddaf55527bb0dc1221e993bd466ef - can you please remove these in a new commit?

BalintKoczor commented:
I have redesigned the code. Now the function calcInnerProduct(...) is overloaded to not only accept state vectors, but density matrices as well. If the arguments are density matrices then their Hilbert-Schmidt scalar product is calculated. The code is currently implemented for CPU (both local and distributed) but not for GPU.

BalintKoczor commented:
....and removed the accidentally committed cmake build files in c474a78631da721f4aa0dc99e3c99ac35f69c810

TysonRayJones commented:
Keep in mind your commit messages appear in context _here_, but once merged, these commits won't be grouped (hence a commit like '_fixed validation_' can appear nondescript). If you approve, we can squash this merge to one commit like 
```added density-matrix calcInnerProduct (CPU)```

TysonRayJones commented:
(also just a reminder that after PR #179, the author list must be updated to:

```
@author Tyson Jones (state vectors)
@author Balint Koczor (density matrices, CPU)
```

BalintKoczor commented:
I have now redesigned the code: I removed all changes to the function calcInnerProduct(), i.e., it only accepts state vectors. The issue with the author list from you previous comment is now fixed.

I added a new function calcDensityInnerProduct(...) that accepts both density matrices and state vectors and the real number it returns for pure states (|<psi|phi>|^2) is identical regardless of whether a pure state is  given as a density matrix or as a state vector.

If all looks good I can squash all commits into one.

TysonRayJones commented on QuEST/src/QuEST.c:
```diff
@@ -834,6 +834,18 @@ Complex calcInnerProduct(Qureg bra, Qureg ket) {
 }
 
+qreal calcDensityInnerProduct(Qureg bra, Qureg ket) {
+    validateMatchingQuregTypes(bra, ket,  __func__);
```
Since this function has ***Density*** in the title, I think it should only accept density matrices. 
I.e. this line should become two:
```C
validateDensityMatrQureg(bra, __func__);
validateDensityMatrQureg(ket, __func__);
```
It's not even performing the inner product when given state-vectors, really!

BalintKoczor replied:
Addressed this in acf7f95214544204939c8267381218a72ec98096.

TysonRayJones commented on QuEST/src/QuEST.c:
```diff
@@ -834,6 +834,18 @@ Complex calcInnerProduct(Qureg bra, Qureg ket) {
     return statevec_calcInnerProduct(bra, ket);
 }
 
+qreal calcDensityInnerProduct(Qureg bra, Qureg ket) {
```
Somewhat pedantic, but with other change (accepting only density matrices), `bra` and `ket` (implying statevectors) would more meaningfully be called `rho` and `omega` (or something conventional for density matrices)

BalintKoczor replied:
Addressed this in acf7f95214544204939c8267381218a72ec98096.

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,43 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
  */
 Complex calcInnerProduct(Qureg bra, Qureg ket);
 
+/** If both input states \p a and \p b are equal-size density matrices
```
Doc should be consistent between argument names (currently `bra` and `ket`, though should become `rho` and `omega`) and symbols in the maths (currently `a` and `b`).

BalintKoczor replied:
Addressed this in acf7f95214544204939c8267381218a72ec98096.

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,43 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
 
+/** If both input states \p a and \p b are equal-size density matrices
+ * then computes the Hilbert-Schmidt scalar product between \p a and \p b, 
+ * which is equialent to the Frobenius inner product of matrices.
```
typo: equialent

BalintKoczor replied:
Addressed this in acf7f95214544204939c8267381218a72ec98096.

TysonRayJones replied:
Haha actually you missed it! :)

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,43 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ * which is equialent to the Frobenius inner product of matrices.
+ * That is, we define the Hilbert-Schmidt scalar product
+ * \f[
+    (( a, b ))_F = ((a, b))_HS := \text{Tr}[ a^\dagger b ] 
```
Order of `(( a, b ))_F = ((a, b))_HS` seem backward, but I'd just remove `(( a, b ))_F`; it provides no more info than is in the previous sentence.

BalintKoczor replied:
Addressed this in acf7f95214544204939c8267381218a72ec98096.

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,43 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ * \f]
+ * This is equivalent to the sum of products of matrix elemets, i.e.
+ * \f[
+    ((a, b))_HS = \sum\limits_i \sum\limits_j  conj(a_{ij}) b_{ij}
```
`((a, b))_HS` should become `((a, b))_{HS}` since `S` currently isn't subscripting

BalintKoczor replied:
Addressed this in acf7f95214544204939c8267381218a72ec98096.

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,43 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ * \f]
+ * This is equivalent to the sum of products of matrix elemets, i.e.
+ * \f[
+    ((a, b))_HS = \sum\limits_i \sum\limits_j  conj(a_{ij}) b_{ij}
```
`conj(a_{ij})` should become `a_{ij}^*` to be consistent with `a^\dagger` used in first formula.

BalintKoczor replied:
Addressed this in acf7f95214544204939c8267381218a72ec98096.

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,43 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ * @throws exitWithError
+ *      if \p bra and \p ket or \p a and \p b have mismatching dimensions
+ *      or mismatching qureg types.
+ * @author Balint Koczor
```
should be
```C
@author Balint Koczor (CPU)
```
unless you write the GPU code!

BalintKoczor replied:
Added (CPU) in acf7f95214544204939c8267381218a72ec98096 for now.

TysonRayJones commented on CMakeLists.txt:
```diff
@@ -12,7 +12,7 @@ project(QuEST_Project)
 # -----------------------------------------------------------------------------
 
-set(USER_SOURCE  "tutorial_example.c"  CACHE STRING "Source to build with QuEST library")
+set(USER_SOURCE  "examples/tutorial_example.c"  CACHE STRING "Source to build with QuEST library")
```
Doing this invalidates the current README, which includes a copy:
```bash
cp examples/tutorial_example .
mkdir build
cd build
cmake ..
make
```
However, I prefer your method here. Can you remove the `cp examples/tutorial_example .` line from the README?

BalintKoczor replied:
Removed line in acf7f95214544204939c8267381218a72ec98096.

BalintKoczor commented:
Thanks for the comments!
calcDensityInnerProduct(...) now only accepts density matrices. I updated the documentation accordingly and switched to the notation `\rho_1` and `\rho_2` as arguments.

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,41 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
 
+/** Computes the Hilbert-Schmidt scalar product
+ * (which is equialent to the Frobenius inner product of matrices) 
+ * of two density matrices \p \rho_1 and \p \rho_2 of equivalent size.
```
The Doxygen args show as `rho1` and `rho2`, so the references in the doc should be `\p rho1` (rather than `\p \rho_1` which isn't a recognised argument).

TysonRayJones replied:
I'll point out all the other places `\p \rho_*` should be changed below so minimise change of missing any

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,41 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
 Complex calcInnerProduct(Qureg bra, Qureg ket);
 
+/** Computes the Hilbert-Schmidt scalar product
+ * (which is equialent to the Frobenius inner product of matrices) 
```
equivalent typo here!

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,41 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ * \f[
+    ((\rho_1, \rho_2))_{HS} = ((\rho_2, \rho_1))_{HS}
+ * \f]
+ * Also note that if both \p \rho_1 and \p \rho_2 are density matrices of pure states
```
`\p \rho_*` -> `\p rho*`

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,41 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ * \f]
+ *
+ * @ingroup calc
+ * @param[in] \p \rho_1 qureg as a density matrix (to have its values conjugate transposed)
```
`\p \rho_*` -> `\p rho*`

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,41 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ *
+ * @ingroup calc
+ * @param[in] \p \rho_1 qureg as a density matrix (to have its values conjugate transposed)
+ * @param[in] \p \rho_2 qureg as a density matrix
```
`\p \rho_*` -> `\p rho*`

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,41 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ * @param[in] \p \rho_1 qureg as a density matrix (to have its values conjugate transposed)
+ * @param[in] \p \rho_2 qureg as a density matrix
+ * @returns the real Hilbert-Schmidt scalar product of density matrices
+            \p \rho_1 and \p \rho_2 (assuming Hermiticity)
```
`\p \rho_*` -> `\p rho*`

TysonRayJones commented on QuEST/include/QuEST.h:
```diff
@@ -1662,6 +1662,41 @@ int measureWithStats(Qureg qureg, int measureQubit, qreal *outcomeProb);
+ * @returns the real Hilbert-Schmidt scalar product of density matrices
+            \p \rho_1 and \p \rho_2 (assuming Hermiticity)
+ * @throws exitWithError
+ *      if \p \rho_1 and \p \rho_2 are not density matrices or
```
`\p \rho_*` -> `\p rho*`
